### PR TITLE
Generalize Frechet

### PR DESCRIPTION
**Pull request by t4c1**
## Summary

Generalizes Frechet distribution related functions to accept Eigen expressions.

## Tests

Tested using expression testing framework.

## Side Effects
None

## Release notes

Generalized Frechet distribution related functions to accept Eigen expressions.

## Checklist

- [x] Math issue #1470

- [x] Copyright holder: Tadej Ciglarič

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: stan/math/prim/prob/frechet_lccdf.hpp [hunk 1]
```diff
@@ -24,44 +26,57 @@ return_type_t<T_y, T_shape, T_scale> frechet_lccdf(const T_y& y,
...
+                                          + !is_constant_all<T_shape>::value
+                                          + !is_constant_all<T_scale>::value
+                                      >= 2>(pow_n / (1.0 / exp_n - 1));
     if (!is_constant_all<T_y>::value) {
-      ops_partials.edge1_.partials_[n] -= alpha_dbl / y_dbl * rep_deriv;
+      ops_partials.edge1_.partials_ = -alpha_val / y_val * rep_deriv;
     }
     if (!is_constant_all<T_shape>::value) {
-      ops_partials.edge2_.partials_[n] -= log(y_dbl / sigma_dbl) * rep_deriv;
+      ops_partials.edge2_.partials_ = -log(y_val / sigma_val) * rep_deriv;
```
Comment (bbbales2) [inline stan/math/prim/prob/frechet_lccdf.hpp:76]: `log(x / y)` -> `log(x) - log(y)`

File: stan/math/prim/prob/frechet_cdf.hpp [hunk 2]
```diff
@@ -25,60 +26,59 @@ return_type_t<T_y, T_shape, T_scale> frechet_cdf(const T_y& y,
...
-    for (size_t n = 0; n < stan::math::size(y); ++n) {
-      ops_partials.edge1_.partials_[n] *= cdf;
-    }
+    ops_partials.edge1_.partials_ = cdf * pow_n * alpha_val / y_val;
   }
   if (!is_constant_all<T_shape>::value) {
-    for (size_t n = 0; n < stan::math::size(alpha); ++n) {
-      ops_partials.edge2_.partials_[n] *= cdf;
-    }
+    ops_partials.edge2_.partials_ = cdf * pow_n * log(y_val / sigma_val);
```
Comment (bbbales2) [inline stan/math/prim/prob/frechet_cdf.hpp:78]: `log(x / y)` -> `log(x) - log(y)`

**Review by bbbales2 — CHANGES_REQUESTED** [review]
I guess just replace the logs and this is good.

File: stan/math/prim/prob/frechet_lcdf.hpp [hunk 3]
```diff
@@ -23,41 +24,50 @@ return_type_t<T_y, T_shape, T_scale> frechet_lcdf(const T_y& y,
...
-      ops_partials.edge2_.partials_[n] += pow_n * log(y_dbl / sigma_dbl);
-    }
-    if (!is_constant_all<T_scale>::value) {
-      ops_partials.edge3_.partials_[n] -= pow_n * alpha_dbl / sigma_dbl;
-    }
+  if (!is_constant_all<T_y>::value) {
+    ops_partials.edge1_.partials_ = pow_n * alpha_val / y_val;
+  }
+  if (!is_constant_all<T_shape>::value) {
+    ops_partials.edge2_.partials_ = pow_n * log(y_val / sigma_val);
```
[the remark this answers is not in the document]
Comment (t4c1) [reply stan/math/prim/prob/frechet_lcdf.hpp]: Performance-wise avoiding `log`s is better. However, from numerical stability point of view you are probably right.

**Review by bbbales2 — APPROVED** [review]

**Pull request merged**